### PR TITLE
Integration lifecycle model proposal

### PR DESCRIPTION
**Pull request by rhuss**
This WIP is published very early to gather some early feedback about the direction. The idea is to have a reconcilation model like for Kubernetes, but specific for integrations.

It doesn't yet contain ideas about versioning and how to include this into the state model.
Actual the most important parts are the static state model and the introduction of the two states used to reconcile an integration: currentState and targetState.

"Pending" actually is not a static state of the intergration but the state of the reconciliation process. However on the UI it might be intermixed with the other states or might be presented by some other means.

### CODE REVIEW

File: proposals/044-integration-model.md [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+* The user must always be able to see and examine the currently active integration along with its version number.
+
+_Attention: THIS IS WORK-IN-PROGRESS. The propsoal is PRed early to collect early feedback_
+
+
+## Integration state model
+
+Technical an Integration is the domain object describing a Camel Route. 
+It is managed by `syndesis-rest` and `syndesis-ui`. 
+During design time this object only lives in this domain.
```
Comment (gashcrumb) [inline proposals/044-integration-model.md]: Question on this one, currently as an integration is created or changed in the UI those changes are only client-side *until* the user decides to actually save the integration.  However the original intent was that changes to the integration would actually be persisted to the backend in some fashion as they're made by the citizen user.

There's pros and cons to each approach, should we figure out which one to use as part of this proposal?

File: proposals/044-integration-model.md [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+| Inactive | Integration is deployed but is not running|
+| Undeployed | An integration has been undeployed | 
+| Error | The integration is deployed but in an error state |
+
+### Draft
+
+Right after an integration is created it enters the _Draft_ state. This means 
+the integration is not yet deployed. An integration leaves the draft state either by being published or deleted (in which case the whole integration is removed)
+
+After an activation there is no way back to the _Draft_ state.
```
Comment (gashcrumb) [inline proposals/044-integration-model.md]: What does having a `draft` state give us?

Would this be an indicator of an 'unfinished' integration for example?

File: proposals/044-integration-model.md [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+
+### Active
+
+An _active_ integration is a running integration. This state can be either created from a _Draft_ state when the integration gets published of from an _Inactive_ state when a suspended integration is started again.
```
Comment (gashcrumb) [inline proposals/044-integration-model.md]: nit:
> published of from

should be

> published from

File: proposals/044-integration-model.md [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+
+### Inactive
+
+An _inactive_ integration is an integration which was previously active and has been suspended by some action. An _inactive_ integration can be resumed or undeployed
+
+### Undeployed
+
+An integration which was published is never deleted but can be undeployed. An undeployed integration can be re-deployed (and activated) if the user chose the version of this integration to be reactivated.
```
Comment (gashcrumb) [inline proposals/044-integration-model.md]: Would `undeployed` integrations show up in the UI anywhere?  How would the user actually totally delete an integration?

File: proposals/044-integration-model.md [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+
+### Error
+
+The error state indicates that the integration is not running. This state can be entered consciously but only as a side effect. An integration can enter the error state any time.
```
Comment (gashcrumb) [inline proposals/044-integration-model.md]: Is it possible for an integration to be running but have a transient error?  Or let's say one connection of the integration winds up having a stale token, the integration is actually running, but is it processing messages (maybe) or do we plan on forcing the integration to just up and die on errors like this?

File: proposals/044-integration-model.md [hunk 6]
```diff
@@ -0,0 +1,203 @@
...
+
+`Error` can never be a `targetState` but it can be a `currentState`. The backend controllers try to get out of the error state but might give up at some point. 
+
+Every state change (current or target) should be tracked. 
+
+## Operations
+
+The following operations change the `targetState`. The `targetState` could be updated by simply changing the Integration's field, but it is recommended to use a more semantic rich REST API call for perfoming this actions (i.e. having specific endpoint which can be used by the UI to trigger the targetState change).
+
+* **Publish** sets the target state to _Active_. This operation can be called from the states:
```
Comment (gashcrumb) [inline proposals/044-integration-model.md]: What if you want to save but not update the running integration?  Or do we not support that?

File: proposals/044-integration-model.md [hunk 7]
```diff
@@ -0,0 +1,203 @@
...
+
+Each integration has a unique name within a certain context (which currently is the whole installation). 
+Multiple versions of an integration can exist. 
+A _active_ integration is immutable.
+So in order to update an active integration (i.e. an integration which is in state _Active_) a new version needs to be created.
+Changing an _active_ integration in the UI causes a new integration to be created, with the following characteristics:
+* Its cloned from the _active_ integration
+* The state of the clone is set to _Draft_
+* The `parentVersion` is set to the version of the active integration it is cloned from
+* The `version` is set to undefined.
```
Comment (gashcrumb) [inline proposals/044-integration-model.md]: This might answer my questions a bit on 'saving' :-)

Comment (gashcrumb) [reply proposals/044-integration-model.md]: Maybe actually this needs to be earlier in the document or referred to in the text leading up to this section?

File: proposals/044-integration-model.md [hunk 8]
```diff
@@ -0,0 +1,203 @@
...
+    
+    // The version of the integration which was the origin of this integration
+    // 0 if this is a new integration
+    int parentVersion;
+    
+    // Version of this integratin
+    int version;
+  
+    // Message describing the state further (e.g. error message)
+    String stateMessage;
```
Comment (gashcrumb) [inline proposals/044-integration-model.md]: I feel like just a string for this may not be that great for communicating problems to the UI and the user if it's used for error messages too, WDYT?

File: proposals/044-integration-model.md [hunk 9]
```diff
@@ -0,0 +1,203 @@
...
+
+## Data model
+
+The following data model only highlights the parts which are specific to the lifecycle model:
+
+```java
+public class Integration {
+    
+    // Unique name of the integration
+    String name;
```
Comment (gashcrumb) [inline proposals/044-integration-model.md]: Should we maybe decouple the user-entered name from the git repo?  People will want to use names like "My awesome message processor".   Would we have a reference to the git repo URL at this level?

File: proposals/044-integration-model.md [hunk 10]
```diff
@@ -0,0 +1,203 @@
...
+    
+    // Current state of the integration
+    IntegrationState currentState;
+    
+    // Target state of the integration    
+    IntegrationState targetState;
```
Comment (gashcrumb) [inline proposals/044-integration-model.md]: Using the same enum is less ideal, would it make sense to have a narrowed enum type for the target state?  As it doesn't make sense for tools to set this to `Error`.

File: proposals/044-integration-model.md [hunk 11]
```diff
@@ -0,0 +1,203 @@
...
+    
+    // The version of the integration which was the origin of this integration
+    // 0 if this is a new integration
+    int parentVersion;
```
Comment (gashcrumb) [inline proposals/044-integration-model.md]: Would commit IDs or tags be better for these version fields?

File: proposals/044-integration-model.md [hunk 12]
```diff
@@ -0,0 +1,203 @@
...
+
+## Deployment
+
+Deployment goes when an integration with `currentState == Draft` set `targetState = Active`. 
+Several actions will be performed by the controller:
+
+* A GitHub repository will be created if this is the first version of an integration (parentVersion == 0). 
+* If it is an new version of an existing integration (e.g. because a user changed something), a new Git branch is created from the old version's branch. This branch is named after the version.
+* The runtime files are created and commit to the Git repository
+* If this is an update of an existing integration, a currently running or suspended integration (`currentState == Active or Inactive`) is set to `targetState = Undeployed` (which triggers an undeployment on behalf of the controller)
```
Comment (gashcrumb) [inline proposals/044-integration-model.md]: This kinda seems then like you can't persist changes without publishing them right away, i.e. what if the user wants to make some updates, persist them but not make those changes active right away?

File: proposals/044-integration-model.md [hunk 13]
```diff
@@ -0,0 +1,203 @@
...
+* The resource descriptors are created for this integration and applied to OpenShift.
+
+Open Question: For an update should an exisiting `DeploymentConfig` be reused (and e.g. doing a rolling update) or should we create a new `DeploymentConfig` ?
+
+## UI changes and updates
+
+Within the UI the state of an integration should be visualized. 
+It is suggested that when the reconciliation state is `Ok` the _currentState_ should be shown and when the reconciliation state is `Pending` then "Pending" should be shown (possibly whith the `targetState` as a tool tip info). Alternatively, when a reconciliation process the label could be more specific about the action just performing ("Deploying", "Suspending", "Resuming", "Undeploying") which can be infered from `currentState` and `targetState`
```
Comment (gashcrumb) [inline proposals/044-integration-model.md]: Yeah, if we maybe have the reconciliation state as a separate field on the integration that'd make the logic in the UI easier.  Most likely when it's `Pending` then we'll show the `currentState` using the appropriate visualization.  I guess though the reconciliation state attribute would have a third state, `Error` so we can show whatever error information is available.

File: proposals/044-integration-model.md (hunk 1 above)
Comment (rhuss) [reply proposals/044-integration-model.md]: This section is a bit unclear. Actually I mean it lives only in the DB but there is not already an associated runtime (pod with camel routs) created.

I would distinguish between "Save" (store it in the DB, lightweight operation) and "Publish" (create the runtime deployment, heavyweight operation).

A "Publish" simply means then to change the `targetState` from "Draft" to "Active" in the DB stored integration. A background controller in syndesis-rest would then pick this up and create the pods. 

So imo the UI is still free to save either explicitely or implicitely for the 'save' operation. Only the "Publish" operation should be a dedicated step (== button).

File: proposals/044-integration-model.md (hunk 2 above)
Comment (rhuss) [reply proposals/044-integration-model.md]: E.g. in _Draft_ the integration needs not to validate it simply a state for 'work-in-progress'. The most important motivation for not allowing it to go back to _Draft_ once out of the state is for a simplified operational mode as the integration then becomes immutable afterwards. When changing such an integration, a new Integration object with  the same name, but another version will be created, again in state _Draft_.

Yes, unfinished integrations are in state _Draft_.

File: proposals/044-integration-model.md (hunk 4 above)
Comment (rhuss) [reply proposals/044-integration-model.md]: That's a very good question which we still need to discuss with @kcbabo . I could imagine to have some kind of "History" page for all integration with different versions. Since would contain all "undeployed" ones, at most one active integration and at most one integration in "Draft" (wip, this draft integration would become active when "Pubished", the active integration would get undeployed).

File: proposals/044-integration-model.md (hunk 5 above)
Comment (rhuss) [reply proposals/044-integration-model.md]: If possible I'd like to keep a single _Error_ state to keep it simple. This state can be left to _Active_ if the error is transient and can get fixed. Along with this state goes a `stateDescription` which can hold detailed information about the error, maybe with some hint how to fix it.

"Error" is not necessarily a final state.

File: proposals/044-integration-model.md (hunk 6 above)
Comment (rhuss) [reply proposals/044-integration-model.md]: If you save but not "publish" nothing will happen on the running integration.

When changing an integration the following steps and checks are performed. 
Please keep in mind that all integration objects with the same name belong together, they are just variants in different versions.

* If there is already an integration with this name in state _Draft_, update this integration.
* If there is no integration in _Draft_ but one in _Active_, the active integration is cloned to a new integration with the same name and in state _Draft_. This integration then can be changed freely and saved to the DB.
* When the changed integration gets published, it's `targetState` is set from "Draft" to "Active". 
* A controller in syndesis-rest detects, that integration with `currentState == Draft` and `targetStat == Active`. The controller then performs these operations in the background:
  - For an integration with `currentState == Active|Inactive` set `targetState = Undeployed`
  - Create / Update the GitHub repo
  - Trigger an OpenShift build  

(i added this to the document)

File: proposals/044-integration-model.md (hunk 7 above)
Comment (rhuss) [reply proposals/044-integration-model.md]: Agreed, I'm going to rearrange the document a bit to clarify the difference between "Save" and "Publish".

File: proposals/044-integration-model.md (hunk 8 above)
Comment (rhuss) [reply proposals/044-integration-model.md]: Actually this class describes the domain model which is not necessarily that what arrives at the UI. I agree that we should strictly decouple user friendly message with technical error message. Any idea how we come from an integration error to a nice UI message ?

File: proposals/044-integration-model.md (hunk 9 above)
Comment (rhuss) [reply proposals/044-integration-model.md]: Good point, also when people want to rename an integration (we might then have to rename all integration instances). Maybe a surrogate, non-visible name would make sense here ?

Actually we could also model the Integration in two levels:

A single `Integration` which contains one or more `IntegrationInstance`s (with at most one active / inactive and most one in draft state).

The difference would be that this is a more explicit grouping instead of an implicit, label-like grouping via the name.

Any opinions what would be better ?

I just added an alternative model to the document.

Comment (rhuss) [reply proposals/044-integration-model.md]: Jus thinking a bit about this, I guess an explicit grouping with two objects makes more sense wrt to consistency. E.g. when changing a name, the repo or any other integration global property, then every such `Integration` object would need to be updated (if grouped via name or any other label like prop)

File: proposals/044-integration-model.md (hunk 10 above)
Comment (rhuss) [reply proposals/044-integration-model.md]: An internal logic (like a monitoring component which uses custom health checks) could set this to `Error` if it want to mark the integration as invalid. 

For the interaction with the UI I think this 'currentState' and 'targetState' should not be exposed to the UI. Its the Rest endpoint which will set the states according to the required logic (e.g. the UI uses an endpoint "publish" and the Rest endpoint sets the targetState to "Active"). 

Of course we could also expose the reconciliation mechanism over the Rest endpoints, too (e.g. that the UI sets the targetState and simply saves the integration). In that case the endpoint could then validate for a valid target state (e.g. not to set the targetState to 'Error' from the UI.

tl;dr - I would like to keep both the same enum so that they are easier to compare and also the targetState could be in fact any state, too, when set internally.

File: proposals/044-integration-model.md (hunk 12 above)
Comment (rhuss) [reply proposals/044-integration-model.md]: Publishing is a deliberate action. Before you publish you can save everything for the integration when it is in _Draft_. So the user can change everyhting without affecting the running integration. It's only after a "Publish" the machinery starts. 

Actually I'm not sure anymore about whether we should set the targetState to "Undeployed" actively. It would be better if the update of the currentState happens in this case when the OpenShift deployment undeploys the current integration in the course of deploying the new one.

I added the following question to the doc:

Question: It would be awesome if we could change the state for an _Active_ integration to _Undeployed_ by watching OpenShift for update deployments so that we can mark the old integration as _Undeployed_ and the new as _Active_. Is this possible with OpenShift ? How to correlate OpenShift deployments with integration versions ?

File: proposals/044-integration-model.md (hunk 9 above)
Comment (gashcrumb) [reply proposals/044-integration-model.md]: Yeah, `Integration` containing `IntegrationInstances` makes sense to me, and I definitely think a system-generated ID of some sort would be good.

File: proposals/044-integration-model.md (hunk 7 above)
Comment (gashcrumb) [reply proposals/044-integration-model.md]: Awesome, and tx for the clarification.  Definitely would like to migrate the UI to updating the integration as the user makes changes so this is all good stuff.

File: proposals/044-integration-model.md (hunk 4 above)
Comment (gashcrumb) [reply proposals/044-integration-model.md]: +1 we definitely could use an integration history page.

File: proposals/044-integration-model.md (hunk 8 above)
Comment (gashcrumb) [reply proposals/044-integration-model.md]: Gotcha, thanks for the clarification.  I guess currently if we're talking about the user looking at the integration list we'd be showing an error icon and most likely this state message will be reflected in the tooltip.  So that might do for a starting point, I mean, we're not expecting the citizen user to go and browse the logs for the pod, instead they'll be asking a more technical user to investigate.

**Comment by rhuss** [comment]
@gashcrumb thanks a lot for all your comments !

I reworked the document, to make it hopefully a bit clearer. Also introduced `IntegrationRevision` which describes a specific version of an integration. An `Integration` is the bracket around all `IntegrationRevisions` and add some global, mutable information.

**Comment by gashcrumb** [comment]
Of course!  Lemme re-read it now with your updates and see if I have any more to add...

File: proposals/044-integration-model.md [hunk 14]
```diff
@@ -0,0 +1,271 @@
...
+
+Once an integration revision is published and transfered to the state _Active_ it becomes **immutable** and can not be changed afterwards (except for state related properties).
+
+`IntegrationRevisions` can be in one of the following state:
+
+![Integration state model](images/integration-state.png "Integration State Model")
+
+| State       | Description |
+| ----------- |------------ |
+| Draft | Initial state of an integration. The `IntegrationRevision` is not yet deployed |
```
Comment (gashcrumb) [inline proposals/044-integration-model.md]: Maybe should be:

> Initial state of an `IntegrationRevision`?

File: proposals/044-integration-model.md [hunk 15]
```diff
@@ -0,0 +1,271 @@
...
+An `Integration` is the domain object describing Camel Routes used for performing integration tasks. 
+Each `Integration` has properties which can be changed by a user from the UI like its name, so its mutable.
+
+The actual logic like the connectors used and the configuration is encapsulated in an `IntegrationRevision`. 
+Multiple `IntegrationRevision` can be connected to a single `Integration`.
+Each `IntegrationRevision` represents a certain version of the defined integration and can be in different states.
+
+Both objects are managed by `syndesis-rest` and `syndesis-ui`. 
+
+When a user creates an integration, an overall `Integration` object is created as well as an `IntegrationRevision` which starts its life in the state _Draft_.
```
Comment (gashcrumb) [inline proposals/044-integration-model.md]: I guess `IntegrationRevision` would always be created in the `Draft` state.

File: proposals/044-integration-model.md [hunk 16]
```diff
@@ -0,0 +1,271 @@
...
+![Integration state model](images/integration-state.png "Integration State Model")
+
+| State       | Description |
+| ----------- |------------ |
+| Draft | Initial state of an integration. The `IntegrationRevision` is not yet deployed |
+| Active | `IntegrationRevision` is deployed and running |
+| Inactive | `IntegrationRevision` is deployed but is not running|
+| Undeployed | An `IntegrationRevision` has been undeployed | 
+| Error | The `IntegrationRevision` is deployed but in an error state |
+
```
Comment (gashcrumb) [inline proposals/044-integration-model.md:59]: Tricky bit with the state here, `Integration` and `IntegrationRevision`, the `Integration`'s state is likely going to always be the most recent `IntegrationRevision`'s state _except_ when the most recent `IntegrationRevision` is in Draft.

File: proposals/044-integration-model.md [hunk 17]
```diff
@@ -0,0 +1,271 @@
...
+`IntegrationRevisions` can be in one of the following state:
+
+![Integration state model](images/integration-state.png "Integration State Model")
+
+| State       | Description |
+| ----------- |------------ |
+| Draft | Initial state of an integration. The `IntegrationRevision` is not yet deployed |
+| Active | `IntegrationRevision` is deployed and running |
+| Inactive | `IntegrationRevision` is deployed but is not running|
+| Undeployed | An `IntegrationRevision` has been undeployed |
```
Comment (gashcrumb) [inline proposals/044-integration-model.md]: Would this maybe be:

> All `IntegrationRevisions` have been undeployed

File: proposals/044-integration-model.md (hunk 14 above)
Comment (rhuss) [reply proposals/044-integration-model.md]: gotcha

File: proposals/044-integration-model.md (hunk 17 above)
Comment (rhuss) [reply proposals/044-integration-model.md]: It's the individual IntegrationRevision which gets undeployed. Another revision can be still active.

The idea is:

* 0 or 1 revision in 'Active' or 'Inactive'
* N in 'Undeployed'
* 0 or 1 in Draft

File: proposals/044-integration-model.md (hunk 16 above)
Comment (rhuss) [reply proposals/044-integration-model.md:59]: Actually, the idea is to have no dedicated state for an Integration, but for individual IntegrationRevisions. The state of the overall Integration is then inferred from the revision's state (but for the UI this overall state is then already provided).

But you are right with respect to the `Error` state, it's a tricky one, because one should keep the original state which should be tried to reach again. But whenn the `state` and the `targetState` are the same, entering the `state` with `Error` will still keep the target state which then is tried to be reconciled.

**Comment by rhuss** [comment]
Going to merge it now, so that we can easily start to work on it.
